src/cli/NativeMsgIface.cpp

```diff
@@ -0,0 +1,454 @@
+            [&](PivotModelInstance &model) {
+                service_->pivot(model.model, model.pivot, std::move(myJsonInput.text.toStdString()), callbackLambda, options);
+            }
+        }, *model_);
+    } catch (const std::runtime_error &e) {
+       lockAndWriteJsonHelper(errJson(myID, QString::fromStdString(e.what())));
+       operations_--;
+    }
+}
+
+inline void NativeMsgIface::handleRequest(ListRequest myJsonInput)  {
+    // Fetch remote models if necessary. In this case we report the models via signal
+    if (myJsonInput.includeRemote && models_.getNewModels().isEmpty()) {
+        models_.fetchRemoteModels(myJsonInput.id);
+    } else {
+        QJsonArray modelsJson;
+        for (auto&& model : models_.getInstalledModels()) {
+            modelsJson.append(model.toJson());
+        }
+        for (auto&& model : models_.getNewModels()) {
+            modelsJson.append(model.toJson());
+        }
+        QJsonObject jsonObj {
+            {"success", true},
+            {"id", myJsonInput.id},
+            {"data", modelsJson}
+        };
+        lockAndWriteJsonHelper(QJsonDocument(jsonObj).toJson());
+        operations_--;
+    }
+
+}
+
+inline void NativeMsgIface::handleRequest(DownloadRequest myJsonInput)  {
+    // Connection to store our callback in. Qt6 has single shot connections, but 5 doesn't.
+    QMetaObject::Connection * const connection = new QMetaObject::Connection;
+
+    auto downloadModelLambda = std::function([=](){
+        // Remove the connection right after we fetch model, if the connection is valid
+        if (*connection)
+            QObject::disconnect(*connection);
+
+        // Clean up our callback connection
+        // TODO: abstract this all away in a connectSingleShot() helper?
+        delete connection;
+        operations_--;
+
+        auto model = models_.getModel(myJsonInput.modelID);
+        if (!model) {
+            lockAndWriteJsonHelper(errJson(myJsonInput.id, QString("Model not found")));
+            return;
+        }
+
+        // Model already downloaded?
+        // TODO: same logic as in downloadComplete signal handler
+        if (model->isLocal()) {
+            QJsonObject jsonObj {
+                {"success", true},
+                {"id", myJsonInput.id},
+                {"data", QJsonObject{{"modelID", model->id()}}}
+            };
+            lockAndWriteJsonHelper(QJsonDocument(jsonObj).toJson());
+            return;
+        }
+
+        // Download new model
+        QNetworkReply *reply = network_.downloadFile(model->url, QCryptographicHash::Sha256, model->checksum, QVariant::fromValue(QMap<QString, QVariant>({
+            {"id", myJsonInput.id},
+            {"modelID", model->id()}
+        })));
+
+        // Pass any download progress updates along to the client. Previous implementation
+        // has some logic to limit the amount of progress updates to 6 but I don't
+        // think throttling the number of messages is necessary.
```

Comment:
My reasoning with regards to throttling is that given a very slow network, the client would be spammed forever with incremental progress messages, though i am not sure if in practise this would mean anything much.